File: src/main/java/com/sy/coladay/reservation/ReservationRepositoryListener.java
package com.sy.coladay.reservation;

import static java.lang.String.format;

import com.sy.coladay.security.AuthenticationFacade;
import com.sy.coladay.user.User;
import java.util.function.Predicate;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.rest.core.annotation.HandleBeforeCreate;
import org.springframework.data.rest.core.annotation.HandleBeforeDelete;
import org.springframework.data.rest.core.annotation.RepositoryEventHandler;
import org.springframework.stereotype.Component;

/**
 * Event handler around auto generated reservation repository.
 */
@Component
@RepositoryEventHandler(Reservation.class)
@AllArgsConstructor
@Slf4j
@SuppressWarnings("unused")
class ReservationRepositoryListener {

  private final AuthenticationFacade authenticationHelper;

  /**
   * Augment the reservation with the <code>Principal</code> being authenticated.
   *
   * @param reservation
   *     the reservation to create
   */
  @HandleBeforeCreate
  @SuppressWarnings("unused")
  void handleReservationBeforeCreate(Reservation reservation) {
    final User authenticatedUser = getAuthenticatedUser();
    LOG.debug("Augmenting reservation with organizer {} from security context ",
        authenticatedUser.getName());
    reservation.setOrganizer(authenticatedUser);
  }

  /**
   * Assert that the <code>Principal</code> being authenticated is the reservation owner to allow
   * the reservation deletion.
   *
   * @param reservation
   *     the reservation to delete
   */
  @HandleBeforeDelete
  @SuppressWarnings("unused")
  void handleReservationBeforeDelete(Reservation reservation) {
    final User authenticatedUser = getAuthenticatedUser();
    final Predicate<User> isDeletionAllowed = user -> reservation.getOrganizer().equals(user);
    if (isDeletionAllowed.negate().test(authenticatedUser)) {
      LOG.info("Reservation {} deletion is forbidden for user {}. Reservation can only be deleted "
                   + "by its owner",
          reservation.getId(),
          authenticatedUser
      );
      throw new ReservationCancellationForbiddenException(
          format("User %s is not allowed to delete this reservation %d",
              authenticatedUser.getName(), reservation.getId()));
    }
  }

  private User getAuthenticatedUser() {
    return authenticationHelper.getCurrentUserPrincipal().getUser();
  }
}

File: src/main/kotlin/com/sy/coladay/reservation/ReservationRepositoryListener.kt
package com.sy.coladay.reservation

import com.sy.coladay.security.AuthenticationFacade
import com.sy.coladay.user.User
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.data.rest.core.annotation.HandleBeforeCreate
import org.springframework.data.rest.core.annotation.HandleBeforeDelete
import org.springframework.data.rest.core.annotation.RepositoryEventHandler
import org.springframework.stereotype.Component
import java.util.function.Predicate

/**
 * Event handler around auto generated reservation repository.
 */
@Component
@RepositoryEventHandler
class ReservationRepositoryListener (private val authenticationHelper: AuthenticationFacade) {

    /**
     * Augment the reservation with the `Principal` being authenticated.
     *
     * @param reservation the reservation to create
     */
    @HandleBeforeCreate
    fun handleReservationBeforeCreate(reservation: Reservation) {
        val authenticatedUser = getAuthenticatedUser()
        LOG.debug(
            "Augmenting reservation with organizer {} from security context ",
            authenticatedUser.name
        )
        reservation.organizer = authenticatedUser
    }

    /**
     * Assert that the `Principal` being authenticated is the reservation owner to allow
     * the reservation deletion.
     *
     * @param reservation the reservation to delete
     */
    @HandleBeforeDelete
    fun handleReservationBeforeDelete(reservation: Reservation) {
        val authenticatedUser = getAuthenticatedUser()
        val isDeletionAllowed = Predicate { user: User? -> reservation.organizer.equals(user) }
        if (isDeletionAllowed.negate().test(authenticatedUser)) {
            LOG.info(
                "Reservation {} deletion is forbidden for user {}. Reservation can only be deleted "
                        + "by its owner",
                reservation.id,
                authenticatedUser
            )
            throw ReservationCancellationForbiddenException(
                String.format(
                    "User %s is not allowed to delete this reservation %d",
                    authenticatedUser.name, reservation.id
                )
            )
        }
    }

    private fun getAuthenticatedUser():User {
        return authenticationHelper!!.getCurrentUserPrincipal().user;
    }


    companion object {

        private val LOG: Logger = LoggerFactory.getLogger(ReservationRepositoryListener::class.java)
    }
}